File: 2023/06.29/송기훈/소문난칠공주.java
package baekjoon.gold;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;

public class 소문난칠공주 {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static char[][] matrix = new char[5][5];

    // 우 하 좌 상
    static int[] dY = {0, 1, 0, -1};
    static int[] dX = {1, 0, -1, 0};
    static boolean[] visited = new boolean[25];
    static int answer = 0;


    public static void main(String[] args) throws IOException {
        for (int i = 0; i < 5; i++) {
            matrix[i] = br.readLine().toCharArray();
        }

        dfs(0, 0, new int[7]);

        System.out.println(answer);

    }

    static void dfs(int index, int depth, int[] selected) {
        if (depth == 7) {
            if (check(selected)) {
                answer += 1;
            }
            return;
        }

        for (int i = index; i < 25; i++) {
            if (!visited[i]) {
                selected[depth] = i;
                visited[i] = true;
                dfs(i + 1, depth + 1, selected);
                visited[i] = false;
            }
        }
    }

    static boolean check(int[] selected) {
        int sCount = 0;

        for (int loc : selected) {
            if (matrix[loc / 5][loc % 5] == 'S') {
                sCount += 1;
            }
        }

        if (sCount < 4) {
            return false;
        }

        // BFS로 연결되어 있는지 확인
        ArrayList<Integer> selectedList = new ArrayList<>();
        for (int loc : selected) {
            selectedList.add(loc);
        }
        Queue<Integer> queue = new LinkedList<>();
        queue.offer(selectedList.get(0));
        while (!queue.isEmpty()) {
            int loc = queue.poll();

            for (int i = 0; i < 4; i++) {
                int nY = loc / 5 + dY[i];
                int nX = loc % 5 + dX[i];

                if (nY < 0 || nY > 4 || nX < 0 || nX > 4) {
                    continue;
                }

                Integer value = nY * 5 + nX;
                if (selectedList.contains(value)) {
                    selectedList.remove(value);
                    queue.offer(value);
                }
            }
        }
        return selectedList.isEmpty();
    }

}

File: 2023/06.29/조수연/소문난_칠공주.kt
import java.io.*
import java.util.*

var answer = 0
val visited = BooleanArray(25)
val students = Array(25) { CharArray(5) }
val dy = intArrayOf(-1,1,0,0)
val dx = intArrayOf(0,0,-1,1)

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {

    val bw = BufferedWriter(OutputStreamWriter(System.out))

    repeat(5) { students[it] = readLine().toCharArray() }

    dfs(0, 0, IntArray(7) { -1 })

    bw.write( "$answer")

    bw.flush()
    bw.close()
}

fun dfs(size: Int, start:Int,  members: IntArray) {

    if (size == 7) {
        if (check(members)) answer++
        return
    }

    for (i in start until 25){
        if(!visited[i]){
            visited[i] = true
            members[size] = i
            dfs(size+1,i+1,members)
            visited[i] = false
        }
    }
}

fun check(members: IntArray) : Boolean{
    var yCount = 0
    for (index in members){
        if (students[index / 5][index % 5] == 'Y') yCount++
    }

    if (yCount > 3) return false // S가 적어도 4개이상 있어야한다.

    val queue : Queue<Int> = LinkedList()
    queue.offer(members[0])
    var memberList = members.toMutableList()

    while (queue.isNotEmpty()){
        val location = queue.poll() // 현재 위치
        for (i in 0 until 4){
            val nextX = location % 5 + dx[i]
            val nextY = location / 5 + dy[i]

            if (nextX in 0 ..4 && nextY in 0..4){
                if (memberList.contains(nextY*5+nextX)){ // 이동한 위치에 memberList에 멤버가 있으면 다음 탐색
                    memberList.remove(nextY*5+nextX)
                    queue.offer(nextY*5+nextX)
                }
            }
        }
    }

    if (memberList.isNotEmpty()) return false // 인접한 멤버가 아니면 false
    return true
}